hpx/lcos/detail/future_data.hpp

```diff
@@ -350,15 +351,15 @@ namespace detail
         // allowed
         void handle_on_completed(completed_callback_type && on_completed)
         {
-
-#if defined(HPX_WINDOWS)
-            bool recurse_asynchronously = false;
-#elif defined(HPX_HAVE_THREADS_GET_STACK_POINTER)
-            bool recurse_asynchronously =
+            // We need to run the completion on a new thread if we are on a
+            // non HPX thread.
+            bool recurse_asynchronously = hpx::threads::get_self_ptr() == nullptr;
+#if defined(HPX_HAVE_THREADS_GET_STACK_POINTER)
+            recurse_asynchronously =
                 !this_thread::has_sufficient_stack_space();
```

Comment:
Does `this_thread::has_sufficient_stack_space()` fail if not executed on a HPX thread?